Fix DHO state overwritten at `LoadComplete` if it is already changed

ekrctb opened the pull request:
- Fixes #14126 (see my comment)

Note: frame-unstable seek (used by the spectator) will load many hit objects in one frame, causes a massive lag. But that is separate from the linked issue, as the linked issue is persistent after seek, not just one frame.

ekrctb committed:
Fix DHO state is overwritten to `Idle` on `LoadComplete`

The state may already be changed before `LoadComplete` is called because DHO is already added to the draw hierarchy.

ekrctb committed:
Add test case for DHO state change before load complete

bdach commented on osu.Game/Rulesets/Objects/Drawables/DrawableHitObject.cs:
```diff
@@ -190,7 +190,9 @@ protected override void LoadComplete()
 
-            updateState(ArmedState.Idle, true);
+            // If the state is changed, transforms are already initialized.
+            if (state.Value == ArmedState.Idle)
```
Adding this guard is the same as specifying `force: false` in `updateState()`, is it not? Which makes me wonder why the true was there in the first place and whether doing this won't break something else.

ekrctb replied:
It is the opposite. `force: false` will apply transforms when the current state is *not* `Idle`.

bdach replied:
...which is different from the guard how? I'm confused. The guard is checking for idle, and not applying transforms forcefully if idle...?

bdach replied:
Never mind, you're correct.

frenzibyte replied:
It reads backwards though, the rare case is when state changes before load, but most of the time it wouldn't. An early-return would make it read better:
```csharp
// The state has already been changed, no need to initialize transforms.
if (state.Value != ArmedState.Idle)
    return;

updateState(ArmedState.Idle, true);
```

ekrctb replied:
This change is obsolete.

ekrctb commented:
The test failure of `TestSliderAnimationDisable` is relevant.
This breaks slider in editor. Not sure why, yet.

ekrctb committed:
Always initialize DHO transforms on LoadComplete

With the previous commit, the transform application is skipped when the state is already changed. But it turns out the previous commit breaks slider animation in the standard editor. This is probably due to the transforms are applied before nested hit objects are added.

ekrctb commented:
Although I didn't confirm exactly why the slider was broken, that is probably because the state change is before the nested drawables are added. To not break existing code such as this, I changed the way to solve the original issue.
Now, it is just using the current state instead of `ArmedState.Idle`.

smoogipoo reviewed (approved):
lgtm

smoogipoo committed:
Merge branch 'master' into fix-dho-state-overwrite

smoogipoo committed:
Merge branch 'master' into fix-dho-state-overwrite
